### PR TITLE
[DPE-4703] - feat: external k8s access via NodePort

### PR DESCRIPTION
**Pull request by marcoppenheimer**
NOTE - tests aren't updated. Will be doing at a later stage before this PR gets merged. Getting early review on implementation.

#### `feat: add expose-external config option`
- Is set to `nodeport` by default
- When enabled, units will get a config-changed, triggering the creation of the NodePort K8s services
- Requires `juju trust`
#### `feat: start bootstrap + listener services for each enabled security.protocol`
- Results in services like the following:
    ```
    kafka-0-sasl-ssl           NodePort    10.152.183.28    <none>        29093:30452/TCP   25m   statefulset.kubernetes.io/pod-name=kafka-0
    kafka-0-ssl                NodePort    10.152.183.22    <none>        29094:30723/TCP   25m   statefulset.kubernetes.io/pod-name=kafka-0
    kafka-1-sasl-ssl           NodePort    10.152.183.33    <none>        29093:31302/TCP   26m   statefulset.kubernetes.io/pod-name=kafka-1
    kafka-1-ssl                NodePort    10.152.183.66    <none>        29094:31657/TCP   26m   statefulset.kubernetes.io/pod-name=kafka-1
    kafka-2-sasl-ssl           NodePort    10.152.183.82    <none>        29093:31304/TCP   25m   statefulset.kubernetes.io/pod-name=kafka-2
    kafka-2-ssl                NodePort    10.152.183.176   <none>        29094:31527/TCP   25m   statefulset.kubernetes.io/pod-name=kafka-2
    kafka-sasl-ssl-bootstrap   NodePort    10.152.183.32    <none>        29093:32273/TCP   26m   app.kubernetes.io/name=kafka
    kafka-ssl-bootstrap        NodePort    10.152.183.148   <none>        29094:32729/TCP   26m   app.kubernetes.io/name=kafka
    ```
- On scale-in, the services are cleaned up by K8s' GC when the pod dies
- Per-pod nodeport numbers are randomly allocated by K8s, and persisted to unit-data
#### `feat: add nodeport-offset config option`
- Defaults to `-1`, which Pydantic validator interprets as `None`
- When set, bootstrap nodeport services are offset by that value
#### `feat: request new TLS certs if node changed`
- We check the current cert's SANs, and if it changed, request a new one with updated SANs for that unit's node
- The unit will remove it's current certificate, and eventually dynamically load it when it's ready. This doesn't cause downtime, as the broker was 'down' already during the reschedule
#### `feat: add K8sManager`
- Bunch of methods for getting nodes, pods, services etc
#### `chore: return node-ip:node-port SASL bootstrap-servers in relation-data`
- This is gotten from `state.security_protocol`, which from now on will serve as the main source-of-truth for which supported auth mechanisms are up and running 
#### `chore: add EXTERNAL scoped ports`
- Prefixed with `20000`, following the `_, 1, 2` pattern we have already

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -92,11 +92,14 @@ def __init__(self, *args):
...
         self.tls_manager = TLSManager(
-            state=self.state, workload=self.workload, substrate=self.substrate
+            state=self.state, workload=self.workload, substrate=self.substrate, config=self.config
         )
         self.auth_manager = AuthManager(
             state=self.state, workload=self.workload, kafka_opts=self.config_manager.kafka_opts
         )
+        self.k8s_manager = K8sManager(
+            pod_name=self.state.unit_broker.pod_name, namespace=self.model.name
+        )
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/charm.py:102]: Agreed on shared manager, but I think that's Pedro's plan down the line.
For the `trust` check, it's at the very bottom of this file:
```python
        try:
            self.client.apply(service)
        except ApiError as e:
            if e.status.code == 403:
                logger.error("Could not apply service, application needs `juju trust`")
                return
```

[the remark this answers is not in the document]
Comment (deusebio) [reply src/charm.py:102]: Nice! I also meant to go into blocked status and be very explicit to the user in the juju status log. I'm afraid that if it is only in the debug-log people may not notice

File: src/core/models.py [hunk 2]
```diff
@@ -5,15 +5,25 @@
...
 
-from literals import INTERNAL_USERS, SECRETS_APP, Substrates
+from literals import (
+    INTERNAL_USERS,
+    SECRETS_APP,
+    SECURITY_PROTOCOL_PORTS,
+    AuthMechanism,
+    Substrates,
+)
+from managers.k8s import K8sManager
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/core/models.py:26]: It's a bit messy, I agree, but given that managers are stateless, I wondered if there was an issue initialising it here, as it's tied to a 'broker' unit/pod. I couldn't find a clean way of doing it otherwise. It shouldn't be circular at least.

As I saw it, a 'manager' is a generic business-logic unit that takes some arguments (i.e not `state` or `charm` directly). I don't see much issue in initialising a manager anywhere it's useful? What do you think?

The planned dependency tree was more focused on `charm` > `state` > `events`, with managers executing logic functionally stateless, was that a wrong interpretation?

File: src/charm.py [hunk 3]
```diff
@@ -258,8 +286,9 @@ def _on_config_changed(self, event: EventBase) -> None:
         if zk_jaas_changed or properties_changed:
             self.on[f"{self.restart.name}"].acquire_lock.emit()
 
-        # update client_properties whenever possible
+        # update these whenever possible
         self.config_manager.set_client_properties()
+        self.update_external_services()
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/charm.py:291]: Note that `charm.py` isn't shared between K8s/VM, they're unique for both (pebble-ready differences as an example) so this is fine to have here.

File: src/core/models.py [hunk 4]
```diff
@@ -228,6 +248,55 @@ def truststore_password(self) -> str:
...
+    @cached_property
+    def node_ip(self) -> str:
+        """The IPV4/IPV6 IP address the Node the unit is on.
+
+        K8s-only.
+        """
+        return self.k8s.get_node_ip(node=self.node)
+
+    @cached_property
+    def listener_nodeports(self) -> dict[AuthMechanism, int]:
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/core/models.py:284]: The `Listener` class in `managers/config.py` builds the config properties `listeners`/`advertised.listeners` etc, whereas here, we have a list of just the node-ports for particular auth-mechanism'd listeners. Fair point, maybe we rename this to `auth_mechanism_nodeports` or something, yeah?

File: src/charm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/charm.py:291]: Although I haven't tested it against VM yet, I am fairly confident that it'll copy over as is with minimal changes, all the core logic for VM was retained.

File: src/core/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/core/models.py:26]: Also note, `managers/k8s.py` does not depend on anything from `core`.

File: src/charm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/charm.py:102]: Ah, yeah I think I'll revisit this with more explicit `Status.NO_TRUST` or something when I come back to this for the tests, good point!

File: src/core/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (deusebio) [reply src/core/models.py:26]: yes, I noticed that too. And in an earlier version of the comment I was suggesting to make K8sManager as part of core, since it is - to some extend - just parsing K8s control-plane information, so it does not sound outrageous. However then, there is also the "business logic" how services are created, and that seems a bit more manager stuff. Unfortunately it seems that `bootstrap_server` that is the reason why we need the k8s manager here, uses also the `business_logic` part. 

I don't have a solution straight away, but I felt it was worth pointing out, maybe among you guys, you can come up with a cleaner solution.  

> I don't see much issue in initialising a manager anywhere it's useful? What do you think?

Sort of. I would still respect the dependency chain (core) => (managers) => (events). Also there can be a dependency chain between managers, e.g. `general manager (e.g config) => more-specific-manager (e.g. auth)`, but if in core you import from manager, there is a risk of potential circularity there (maybe you can avoid that, right now, but it may bite back later). So more-specific manager can import from other managers, and events can import from all of the others. But core cannot import events. That would be weird, wouldn't it?

File: src/charm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (deusebio) [reply src/charm.py:291]: Yes, definitely. `charm.py` can be substrate specific. However, right now we have a bit of common logic and copy and paste there too. Actually, I was recently discussing this with Raul as well, as suggesting to try to separate and factor out from `charm.py` the "common" part, such that `charm.py` only has "substrate" specific stuff. 

Note that I mentioned this here since it was within a sort-of "general" hook, the _changed_config one. The substrate specific logic in substrate specific hooks (e.g. pebble ready event) can surely stay!

[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/charm.py:291]: Yeah, I'd like to work on leaning-in to the copied code across both, in `charm.py` and in the integration-tests especially. I'll keep that in mind, but probably not to address in this PR.

File: pyproject.toml [hunk 5]
```diff
@@ -38,7 +38,7 @@ jsonschema = ">=4.10"
 cryptography = ">42.0.0"
 pydantic ="^1.10.7"
 pyyaml = "^6.0.1"
-lightkube = "^0.15.0"
+lightkube = "0.15.0"
```
Comment (Batalex) [inline pyproject.toml:41]: **question**: Did you experience breaking changes in the `.2` and `.3` patch releases? The changelogs are quite sparse, when they exist.
Also, why do we need `lightkube` in the lint group?

File: src/core/cluster.py [hunk 6]
```diff
@@ -167,6 +170,16 @@ def clients(self) -> set[KafkaClient]:
 
     # ---- GENERAL VALUES ----
 
+    @property
+    def bind_address(self) -> IPv4Address | IPv6Address | str:
+        """The network binding address from the peer relation."""
+        bind_address = None
+        if self.peer_relation:
+            if binding := self.model.get_binding(self.peer_relation):
```
Comment (Batalex) [inline src/core/cluster.py:178]: **praise**: Spread the walrus love!

File: src/core/cluster.py [hunk 7]
```diff
@@ -211,7 +252,17 @@ def bootstrap_server(self) -> str:
         if not self.peer_relation:
             return ""
 
-        return ",".join(sorted([f"{broker.host}:{self.port}" for broker in self.brokers]))
+        if self.config.expose_external:  # implicitly checks for k8s in structured_config
+            return self.bootstrap_servers_external
```
Comment (Batalex) [inline src/core/cluster.py:256]: **praise**: Seems like a sensible solution to keeping this file copy-able / past-able between VM and K8s

File: src/core/structured_config.py [hunk 8]
```diff
@@ -224,6 +227,18 @@ def profile_values(cls, value: str) -> str | None:
...
 
+    @validator("expose_external")
+    @classmethod
+    def expose_external_validator(cls, value: str) -> str | None:
+        """Check expose-external config option is only used on Kubernetes charm."""
+        if SUBSTRATE == "vm" and value:
+            raise ValueError("Value not permitted on VM charm")
+
+        if value == "none":
+            return
```
Comment (Batalex) [inline src/core/structured_config.py:238]: **question**: Should we invert those two conditions? With this flow, we do not allow VM to pass `"none"`. Or will we remove the default value in the VM `config.yaml` file, e.g.
```yaml
  expose-external:
    description: "String to determine how to expose the Kafka cluster externally from the Kubernetes cluster. Possible values: 'nodeport', 'none'"
    type: string
    default: 
```

instead of 
```yaml
  expose-external:
    description: "String to determine how to expose the Kafka cluster externally from the Kubernetes cluster. Possible values: 'nodeport', 'none'"
    type: string
    default: "none"
```
?

File: src/core/models.py [hunk 9]
```diff
@@ -155,17 +166,23 @@ def unit_id(self) -> int:
...
-            host = f"{self.unit.name.split('/')[0]}-{self.unit_id}.{self.unit.name.split('/')[0]}-endpoints"
+            addr = f"{self.unit.name.split('/')[0]}-{self.unit_id}.{self.unit.name.split('/')[0]}-endpoints"
 
-        return host
+        return addr
+
+    @property
+    def host(self) -> str:
+        """Return the hostname of a unit."""
+        return self.node_ip or self.internal_address  # node_ip will be empty on  VM charms
```
Comment (Batalex) [inline src/core/models.py:185]: **praise**: There again, what a nice way to keep things consistent

File: src/events/zookeeper.py [hunk 10]
```diff
@@ -83,14 +86,15 @@ def _on_zookeeper_changed(self, event: RelationChangedEvent) -> None:
 
         # attempt re-start of Kafka for all units on zookeeper-changed
         # avoids relying on deferred events elsewhere that may not exist after cluster init
-        if not self.charm.healthy and self.charm.state.cluster.internal_user_credentials:
+        if not self.charm.healthy:
```
Comment (Batalex) [inline src/events/zookeeper.py:89]: **praise**: Nice job finding this redundant conditional

File: src/managers/config.py [hunk 11]
```diff
@@ -45,11 +46,15 @@ class Listener:
...
 
-    def __init__(self, host: str, protocol: AuthProtocol, mechanism: AuthMechanism, scope: Scope):
-        self.protocol: AuthProtocol = protocol
-        self.mechanism: AuthMechanism = mechanism
+    def __init__(
+        self, auth_map: AuthMap, scope: Scope, host: str = "", node_port: int | None = None
+    ):
+        self.auth_map = auth_map
+        self.protocol = auth_map[0]
+        self.mechanism = auth_map[1]
```
Comment (Batalex) [inline src/managers/config.py:54]: **suggestion (non-blocking)**: Maybe we could have made `AuthMap` a named tuple instead of a regular tuple alias for the extra readability

**Review by Batalex — COMMENTED** [review]
Mostly questions

File: .gitignore [hunk 12]
```diff
@@ -10,3 +10,4 @@ __pycache__/
 .idea
 .python-version
 last_written_value
+todo.md
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply .gitignore:13]: Ah honestly this should be removed, I added it ages ago. I'll do that now.

File: config.yaml [hunk 13]
```diff
@@ -90,3 +90,7 @@ options:
     description: 'Level of logging for the different components operated by the charm. Possible values: ERROR, WARNING, INFO, DEBUG'
     type: string
     default: "INFO"
+  expose-external:
+    description: "String to determine how to expose the Kafka cluster externally from the Kubernetes cluster. Possible values: 'nodeport', 'none'"
+    type: string
+    default: "nodeport"
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply config.yaml:96]: I think so yeah, `ingress` specifically.

File: pyproject.toml (hunk 5 above)
Comment (marcoppenheimer) [reply pyproject.toml:41]: I was having huge issues in the typing inference of it, actually opened an issue for it - https://www.github.com/gtsystem/lightkube/issues/63
It 'seems' to be resolved, but we can look at it in another PR.

File: src/core/structured_config.py (hunk 8 above)
Comment (marcoppenheimer) [reply src/core/structured_config.py:238]: Ah shit, VERY good point. I don't have a good idea here. Maybe we specifically change the default value on the VM charm, what do you think?

File: src/managers/config.py (hunk 11 above)
Comment (marcoppenheimer) [reply src/managers/config.py:54]: Good suggestion, will do!

File: src/core/structured_config.py (hunk 8 above)
Comment (Batalex) [reply src/core/structured_config.py:238]: I think it's ok to have different defaults between the two charms right now. But I also think this raises a bigger question: would it be ok to just ignore this config option in VM (and eventually log that we ignore this option, for ~slightly better user XP)?

Because when (if?) we merge the two code bases in one single fit-all charm, this will be a problem

Comment (marcoppenheimer) [reply src/core/structured_config.py:238]: I suppose I could change the validator to also return `none` if SUBSTRATE == "vm". Could do that now, what do you think?

Comment (Batalex) [reply src/core/structured_config.py:238]: I would prefer that, we can always go back to blocking the charm if needed

Comment (deusebio) [reply src/core/structured_config.py:238]: Uhm, I don't think we should expose the config option on VMs given it does not make much sense there: thus, I would remove the expose-external option from config.yaml on VM charms. It is still fine to have that key in the code (such that we can maximise the code sharing) and hard code the attribute in VM on charm.py or somewhere where it is convenient on VM.

Comment (marcoppenheimer) [reply src/core/structured_config.py:238]: We're copying `structured_config`, I can check if it handles just not passing that config option?

File: src/core/cluster.py [hunk 14]
```diff
@@ -35,9 +37,10 @@
 class ClusterState(Object):
     """Collection of global cluster state for the Kafka services."""
 
-    def __init__(self, charm: Framework | Object, substrate: Substrates):
+    def __init__(self, charm: Framework | Object, substrate: Substrates, config: CharmConfig):
         super().__init__(parent=charm, key="charm_state")
         self.substrate: Substrates = substrate
+        self.config = config
```
Comment (zmraul) [inline src/core/cluster.py:43]: **question**: Why not just get config through charm? With the right type it works the same:
```python
class ClusterState(Object):
    def __init__(self, charm: "KafkaCharm", substrate: Substrates):
        charm.config.<this is resolved to CharmConfig values>
```

Comment (marcoppenheimer) [reply src/core/cluster.py:43]: Good point! Have done.

File: src/managers/config.py [hunk 15]
```diff
@@ -9,6 +9,8 @@
 import re
 import textwrap
 
+from lightkube.core.exceptions import ApiError
```
Comment (Batalex) [inline src/managers/config.py:12]: **thought**: I wonder if we should define our own `exceptions.py` with abstractions. When it comes to executing stuff on the VM or container, it's OK to have `except (subprocess.CalledProcessError, ops.pebble.ExecError)` because one of those is from the std lib, and the other is from `ops` which is present in both VM and k8s.

The issue here would be that `lightkube` is a k8s only dependency, giving us a hard time to copy and paste the code between the two charms.
What do you think about having an `exceptions.py` with the following content?
```py
from literals import SUBSTRATE

if SUBSTRATE == "k8s":
    from lightkube.core.exceptions import ApiError

    class K8sApiError(ApiError):
        pass

else:
    class K8sApiError(Exception):
        pass
```

for the copy/pastable version, or a different file altogether like we do we `literals.py`?

File: src/managers/k8s.py [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+        self.short_auth_mechanism_mapping: dict[AuthMechanism, str] = {
+            "SCRAM-SHA-512": "scram",
+            "OAUTHBEARER": "oauth",
+            "SSL": "ssl",
+        }
+
+    @cached_property
+    def client(self) -> Client:
+        """The Lightkube client."""
+        return Client(  # pyright: ignore[reportArgumentType]
```
Comment (Batalex) [inline src/managers/k8s.py:48]: **note**: this type issue from the package is annoying. If I find the motivation, I'll submit a patch to upstream

**Review by Batalex — APPROVED** [review]

File: src/managers/config.py (hunk 15 above)
Comment (zmraul) [reply src/managers/config.py:12]: +1 to go towards a separate file eventually

Comment (marcoppenheimer) [reply src/managers/config.py:12]: Is there any issue with also making `lightkube` a dep on VM? It's not being used beyond the error type. I don't mind making a custom exception, but it doesn't seem to be a big cost.

Comment (deusebio) [reply src/managers/config.py:12]: My preference would be to have a custom error, since in VM we could also drop the `lightkube` dependency. This custom error seems to me to be connected to the manager. I believe it is sane to import in `managers.config` from the `managers.k8s` both the Manager and the custom exceptions related to it. 

Would `manager/k8s.py` be copied and pasted between VM and K8s? Maybe we shouldn't, we could just provide two implementations that in K8s is an actual implementation, while in VM it is a no-op for the manager and the dummy definition for k8s.

File: src/core/cluster.py [hunk 17]
```diff
@@ -26,18 +28,22 @@
...
+
 
 class ClusterState(Object):
     """Collection of global cluster state for the Kafka services."""
 
-    def __init__(self, charm: Framework | Object, substrate: Substrates):
+    def __init__(self, charm: "KafkaCharm", substrate: Substrates):
         super().__init__(parent=charm, key="charm_state")
         self.substrate: Substrates = substrate
+        self.config = charm.config
```
Comment (deusebio) [inline src/core/cluster.py:46]: **huge suggestion** If this is the reason for having the charm in the constructor, why not passing as argument the `config` object? The definition of the config object is still in `core` (in structure_config), so we would not need to break the clean import (here we are importing from charm which is at the end of the dependency tree) for the typing (yes, I understand it is only for the typing, but we really don't need this I believe)

File: src/events/upgrade.py [hunk 18]
```diff
@@ -165,3 +170,26 @@ def _set_rolling_update_partition(self, partition: int) -> None:
...
+            try:
+                internal_user_credentials = self.charm.zookeeper._create_internal_credentials()
+            except (KeyError, RuntimeError, subprocess.CalledProcessError, ExecError) as e:
+                logger.warning(str(e))
+                event.defer()
+                return
+
+            # only set to relation data when all set
+            for username, password in internal_user_credentials:
+                self.charm.state.cluster.update({f"{username}-password": password})
```
[the remark this answers is not in the document]
Comment (deusebio) [reply src/events/upgrade.py:191]: **suggestion** which one will land first? I suggest you guys to decide which one you want to land first and coordinate in order to simplify your each other work. I'll definitely leave this up to you guys. I don't have any strong preference here, maybe just a small nudge: since this features would still require backporting to vm, in the process of backporting may anyway need to tweaking and changes (I reckon), whereas the CC features seems more aligned between VM and K8s, so I would prioritize that one. But again, up to you

File: src/core/cluster.py (hunk 17 above)
Comment (marcoppenheimer) [reply src/core/cluster.py:46]: Lol, I had originally done it that way, see https://github.com/canonical/kafka-k8s-operator/pull/110#discussion_r1696552026

@zmraul - what was your reasoning here?

File: src/managers/config.py (hunk 15 above)
Comment (marcoppenheimer) [reply src/managers/config.py:12]: I think we should copy the file over, just to reduce the number of differences.
Again, is there any issue with a `lightkube` dependency? We'll eventually have it with 'single-charm'.

Comment (Batalex) [reply src/managers/config.py:12]: I guess we could make the point that adding `lightkube` to VM is a step towards unifying the two codebases. I had a look at the package, and it's fairly small anyway, even including the secondary dependencies

File: src/core/cluster.py (hunk 17 above)
Comment (Batalex) [reply src/core/cluster.py:46]: My 2 cents would be to pass the necessary and sufficient dependencies, so, like the original work did. The reasoning would usually be around surface attack, dependency tree, and ease to test stuff (for instance, we would not have to instantiate a complete `KafkaCharm` to run unit tests on `ClusterState`.
That being said, we do not have those type of laser focused unit tests, and it can be convenient to pass `charm` rather than X different attributes.

tl;dr: really tiny +0.01 on passing `config`

Comment (deusebio) [reply src/core/cluster.py:46]: To me it is for both keeping scope confined AND clear dependency tree. It is definitely more than 0.01 for me :joy:

Comment (zmraul) [reply src/core/cluster.py:46]: Keeping the typing as `Framework | Object` is fine. I think that if we find more cases like this, where`cluster` or `core` are in need of accessing things that come from charm in execution time, we are forcing the dependency tree into what we _actually_ need. 
Passing as an explicit argument is a good way to keep scope and dependency, but to me it points to the code not being able to strictly follow the pattern.

Comment (marcoppenheimer) [reply src/core/cluster.py:46]: While I'm fine with either option, I will say that the requirement to avoid passing 'charm' around **was** supposed to make mocking state/config easier with Scenario. For those that have used Scenario, was it any easier than with `harness.charm`?

Comment (Batalex) [reply src/core/cluster.py:46]: Tiny nitpick, `Framework` is a subclass of `Object` so there is probably no need for the union of types. I do not remember any control flow related to that.

> While I'm fine with either option, I will say that the requirement to avoid passing 'charm' around was supposed to make mocking state/config easier with Scenario. For those that have used Scenario, was it any easier than with harness.charm?

Hum, I would say that there is no difference at all, because Scenario entry point [is the charm itself](https://github.com/canonical/ops-scenario/blob/main/scenario/context.py#L234), you don't test an event handler in isolation. Though I could be wrong, but the more I think about it, the more I believe that given sufficient complexity, scenario would work best with simple test charms dedicated to testing one event handler at a time
